[BUG] Flaky test: test_call_tool_async_caller_cancellation_timeout_preserves_cancelled_error

### Checks

- [x] I have updated to the lastest minor and patch version of Strands
- [x] I have checked the documentation and this is not expected behavior
- [x] I have searched [./issues](./issues?q=) and there are no duplicates of my issue

### SDK Language

Python

### Strands Version

`main` @ `0e517d62` (repo checkout, not an installed release)

### Language Runtime Version

Python 3.13

### Operating System

macOS 26.5.2 (Apple silicon, 15 cores, pytest-xdist)

### Installation Method

git clone

### Steps to Reproduce

Run the Python unit suite repeatedly from `strands-py/` on an unmodified `main`:

```bash
hatch test --python 3.13
```

`tests/strands/tools/mcp/test_mcp_client.py::test_call_tool_async_caller_cancellation_timeout_preserves_cancelled_error` fails intermittently — 1 failure in 3 runs here, plus one during a separate `hatch run prepare`. It passes 100% in isolation; it only appears under parallel full-suite load.

To make it deterministic, add one line to `_invoke_on_background_thread` in `src/strands/tools/mcp/mcp_client.py`, between the log at line 1340 and the await at line 1342:

```python
await asyncio.sleep(0.5)
```

It then fails every time, in isolation, in ~3 seconds.

### Expected Behavior

The test passes regardless of machine load or worker scheduling.

### Actual Behavior

```
    assert "did not finish within bounded cancellation cleanup" in caplog.text
>   assert client._background_cleanup_tasks
E   assert set()
```

The log assertion on the preceding line passes — only the `_background_cleanup_tasks` assertion fails.

### Additional Context

The test waits on the log line as a proxy for "the unfinished invocation has been retained", but the client logs that *before* it retains the task:

```python
if pending:
    self._log_debug_with_thread("MCP invocation did not finish within bounded cancellation cleanup")  # 1340
if cancellation_state is not None:
    await self._cancel_tool_call(cancellation_state)                                                  # 1342
if pending:
    self._track_background_cleanup_task(invoke_event)                                                 # 1346
```

`_cancel_tool_call` is not cheap — it creates a cancellation task and waits on it for up to another second. The test exits its poll the instant the log appears and asserts immediately, observing the client mid-sequence. The client runs its own loop on a separate thread, so nothing orders the two.

Since the log assertion passing proves `pending` was truthy, and `pending` isn't reassigned between 1339 and 1345, line 1346 is guaranteed to run — so the only empty window is between the log and 1346.

| Scenario | Result |
| --- | --- |
| Unmodified, full suite | 1 failure in 3 runs (small sample) |
| Unmodified, isolation | 0 failures |
| Unmodified + injected 0.5s delay | **3 / 3 failed** |
| Fixed test + same delay | 5 / 5 passed |
| Fixed test, no delay, full suite | **30 / 30 passed** |

One caveat I could not rule out: `_cancel_tool_call` catches `except Exception`, which does not catch `asyncio.CancelledError`. If one escaped on this path, line 1346 would be skipped and the set would stay empty *permanently* — same symptom, but not fixable by polling in the test. Thirty clean runs suggest this doesn't happen at a rate that matters, but I couldn't prove it unreachable. If the flake recurs after a test-side fix, that's where I'd look.

Introduced four days ago in #3402, which closed the port request #3276. Worth noting the bounded-cleanup machinery is Python-only — `strands-ts/src/mcp/` has no equivalent of `_background_cleanup_tasks` or the bounded cancellation window, since the TypeScript side cancels via `AbortSignal` on a single thread. So there is no TypeScript counterpart carrying the same bug.

### Possible Solution

Wait for the condition being asserted rather than a log line that precedes it:

```python
for _ in range(150):
    if client._background_cleanup_tasks:
        break
    await asyncio.sleep(0.01)
```

The existing log assertion stays; it just stops being the loop's exit condition. Happy to raise a PR.

### Related Issues

#3402, #3276

--- a/strands-py/tests/strands/tools/mcp/test_mcp_client.py
+++ b/strands-py/tests/strands/tools/mcp/test_mcp_client.py
@@ -555,8 +555,12 @@ async def test_call_tool_async_caller_cancellation_timeout_preserves_cancelled_e
                 with pytest.raises(asyncio.CancelledError):
                     await caller
                 assert await asyncio.to_thread(call_cancelled.wait, 1)
+                # The client logs the expired window, then sends the cancellation notification, and
+                # only then retains the unfinished invocation. Waiting on the log alone would
+                # observe the client mid-sequence, so wait for the retained task itself. The
+                # client runs on its own thread, so nothing else orders these two observations.
                 for _ in range(150):
-                    if "did not finish within bounded cancellation cleanup" in caplog.text:
+                    if client._background_cleanup_tasks:
                         break
                     await asyncio.sleep(0.01)
 

